File: URL-Shortner/fastapi-service/app/cache.py
import json
from datetime import datetime, timezone

import redis.asyncio as redis

from app.config import settings

client: redis.Redis | None = None


async def connect() -> None:
    global client
    client = redis.from_url(settings.redis_url, decode_responses=True)


async def disconnect() -> None:
    if client is not None:
        await client.aclose()


def _key(short_code: str) -> str:
    return f"link:{short_code}"


async def get_link(short_code: str) -> dict | None:
    raw = await client.get(_key(short_code))
    if raw is None:
        return None
    data = json.loads(raw)
    if data["expires_at"] is not None:
        data["expires_at"] = datetime.fromisoformat(data["expires_at"])
    return data


async def set_link(short_code: str, long_url: str, expires_at: datetime | None) -> None:
    # Cache only what the redirect hot path needs, not the whole row.
    value = json.dumps(
        {
            "long_url": long_url,
            "expires_at": expires_at.isoformat() if expires_at else None,
        }
    )

    ttl_seconds = settings.default_cache_ttl_seconds
    if expires_at is not None:
        seconds_until_expiry = int((expires_at - datetime.now(timezone.utc)).total_seconds())
        ttl_seconds = max(1, min(ttl_seconds, seconds_until_expiry))

    await client.set(_key(short_code), value, ex=ttl_seconds)

File: URL-Shortner/go-service/internal/cache/cache.go
package cache

import (
	"context"
	"encoding/json"
	"time"

	"github.com/redis/go-redis/v9"
)

type Client struct {
	rdb        *redis.Client
	defaultTTL time.Duration
}

// LinkValue is intentionally minimal - only what the redirect hot path
// needs, not the whole links row.
type LinkValue struct {
	LongURL   string     `json:"long_url"`
	ExpiresAt *time.Time `json:"expires_at"`
}

func New(redisURL string, defaultTTLSeconds int) (*Client, error) {
	opt, err := redis.ParseURL(redisURL)
	if err != nil {
		return nil, err
	}
	return &Client{
		rdb:        redis.NewClient(opt),
		defaultTTL: time.Duration(defaultTTLSeconds) * time.Second,
	}, nil
}

func key(shortCode string) string {
	return "link:" + shortCode
}

func (c *Client) GetLink(ctx context.Context, shortCode string) (*LinkValue, error) {
	raw, err := c.rdb.Get(ctx, key(shortCode)).Result()
	if err == redis.Nil {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	var v LinkValue
	if err := json.Unmarshal([]byte(raw), &v); err != nil {
		return nil, err
	}
	return &v, nil
}

func (c *Client) SetLink(ctx context.Context, shortCode, longURL string, expiresAt *time.Time) error {
	data, err := json.Marshal(LinkValue{LongURL: longURL, ExpiresAt: expiresAt})
	if err != nil {
		return err
	}

	ttl := c.defaultTTL
	if expiresAt != nil {
		if untilExpiry := time.Until(*expiresAt); untilExpiry < ttl {
			ttl = untilExpiry
		}
		if ttl < time.Second {
			ttl = time.Second
		}
	}

	return c.rdb.Set(ctx, key(shortCode), data, ttl).Err()
}
